### PR TITLE
Fix error reporting of popen and pclose for Windows (issue 427)

### PR DESCRIPTION
**Pull request by mikucionisaau**
Attempt to fix [issue 427](https://github.com/alandefreitas/matplotplusplus/issues/427)

### CODE REVIEW

**Comment by alandefreitas** [comment]
Oh... That's interesting.

**Comment by mikucionisaau** [comment]
@alandefreitas I think the core issue is that `popen` is hackish: 
1) OS does a lot of strange things behind the scenes, like recovering the child process handle from `FILE` handle.
2) the previous implementation used some weird (internal?) features I could not figure out how they work. Quite possible that something has changed in the C library, and thus recovering the handle to the child process does not work anymore.

So instead, I rewrote `popen` functionality from scratch (it appears that it is quite common to replace `popen` with custom implementation, even for Linux) and wrapped `FILE` and process handle into C++ classes.

On one hand, one can make C API with a custom data structure (looks ugly, see commit history).
On the other hand, one can even wrap those pipe classes into `std::istream` and `std::ostream`, expose the rich API and hide the process pipe complexity, but the gnuplot backend wrapper does not need that much.

It would be nice to test this PR on a Mac before merging.

File: source/matplot/backend/gnuplot.cpp [hunk 1]
```diff
@@ -323,11 +321,9 @@ namespace matplot::backend {
         static std::string terminal_type;
         const bool dont_know_term_type = terminal_type.empty();
         if (dont_know_term_type) {
-            terminal_type =
-                run_and_get_output("gnuplot -e \"show terminal\" 2>&1");
-            terminal_type = std::regex_replace(
-                terminal_type, std::regex("[^]*terminal type is ([^ ]+)[^]*"),
-                "$1");
```
Comment (mikucionisaau) [inline source/matplot/backend/gnuplot.cpp:330]: @alandefreitas `std::regex` is ridiculously slow, consider `string_view`:
https://quick-bench.com/q/IUXYt_GIlS3VW8Gh4hJg8HLVQj4

Comment (alandefreitas) [reply source/matplot/backend/gnuplot.cpp:330]: Yes. A custom parser would be better here, also in terms of compilation time.

**Comment by alandefreitas** [comment]
Oh... Thanks! This ended up being quite complex. Nice work. Is it good to go? It looks good to me.

CI would check Apple Clang, but it didn't run for some reason. It seems it needs some maintenance anyway because the latest runs broke in https://github.com/alandefreitas/matplotplusplus/actions. That's expected from time to time. It shouldn't block this PR because it's unrelated to it, though.

**Comment by mikucionisaau** [comment]
I just fixed error reporting for MSVC, which is very weird.
Also tested on MacOS with Intel architecture which has no leak sanitizer and address sanitizer seems broken, so I had to disable all sanitizers. The examples seem to work, but I get lots of noise in the output, not sure whom to blame.

Sorry for the big chunk of code.
I think it's ready, but feel free to review and complain :-)

**Pull request merged**